### ISSUE TITLE
Intermittent crash during style rule GC

### ISSUE THREAD

**Issue by jdm** [servo/servo#26129]
Observed during a CI run:
```  
  ▶ CRASH [expected PASS] /css/CSS2/tables/table-anonymous-objects-011.xht
  │ 
  │ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  │ 
  │ 
  │ 
  │ Shutting down the Constellation after generating an output file or exit flag specified
│ Stack trace for thread "LayoutThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }"
  │ stack backtrace:
  │    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
  │    1: core::fmt::write
  │    2: std::io::Write::write_fmt
  │    3: servo::backtrace::print
  │    4: servo::install_crash_handler::handler
  │    5: _sigtramp
  │    6: style::rule_tree::StrongRuleNode::gc
  │    7: <style::rule_tree::RuleTree as core::ops::drop::Drop>::drop
  │    8: core::ptr::drop_in_place
  │    9: core::ptr::drop_in_place
  │   10: layout_thread::LayoutThread::start
  │   11: profile_traits::mem::ProfilerChan::run_with_memory_reporting
  │   12: std::sys_common::backtrace::__rust_begin_short_backtrace
  │   13: core::ops::function::FnOnce::call_once{{vtable.shim}}
  │   14: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  │   15: std::sys::unix::thread::Thread::new::thread_start
  │   16: _pthread_body
  └   17: _pthread_start
```

**Comment by emilio** [comment]
I think this has been caused by the recent changes by @nox. We rely on the rule nodes to be dropped from the same thread once the rule tree is gone.

**Comment by nox** [comment]
Why do they need to be dropped there?

**Comment by nox** [comment]
I'm confused as to how "rule nodes" relate to the thing I made owned by the DOM, namely a `StyleData` value.

**Comment by nox** [comment]
@emilio Is this something that Gecko also does btw? Not dropping style data from the script thread? If not, how do they do?

**Comment by emilio** [comment]
A `StyleData` contains styles, which contain `StrongRuleNode` objects.

> @emilio Is this something that Gecko also does btw? Not dropping style data from the script thread? If not, how do they do?

In gecko script-thread == layout-thread == main-thread, so this issue doesn't arise.

**Comment by nox** [comment]
If something is `Send`, it means it can be dropped from any thread we damn want. What you are telling me is that `Arc<ComputedValues>` cannot actually be dropped from any thread we damn want. That seems like a very important detail that should probably be stated in bold somewhere in the documentation.

**Comment by SimonSapin** [comment]
> We rely on the rule nodes to be dropped from the same thread once the rule tree is gone.

Can you say more about that? Why does it matter which thread runs a destructor? Is TLS involved?

**Comment by emilio** [comment]
> Can you say more about that? Why does it matter which thread runs a destructor? Is TLS involved?

It doesn't matter _which_ thread. It matters that there are no concurrent drops after the stylist / rule tree is dropped. This is because it [invokes `gc()`](https://searchfox.org/mozilla-central/rev/a707541ff423ade0d81cef6488e6ecfa09273886/servo/components/style/rule_tree/mod.rs#1570) which is not thread-safe, effectively.

**Comment by nox** [comment]
Layout 2020 has been designed from day one with the assumption that `Arc<ComputedValues>` could be stored from wherever we want. What do you think of making `StrongRuleNode::drop` threadsafe with the `servo` feature in stylo?

**Comment by emilio** [comment]
Yes, that'd be welcome. Though I think the merging of the layout and script threads (which I read something about on riot) would also fix this effectively.

**Comment by nox** [comment]
It would not fix this, no. If there is something `'static` in an `Arc`, you are not supposed to need to be careful when dropping it. I don't want to have to be careful, this seems at odds with the whole "fearless concurrency" thing.

**Comment by nox** [comment]
I'm totally fine with it being gated on a `servo`-specific feature though, given AFAIK that may introduce a bit more overhead when dropping `StrongRuleNode` after the `RuleTree` is gone.

**Comment by nox** [comment]
To do any of this though, I also need to make the drop impl be non-recursive, in which case I think [this paper](https://hal.inria.fr/hal-02177326v2) can help.

**Comment by nox** [comment]
So I made https://github.com/nox/recycling-tree as a way to think about this without the Stylo-specific parts that were overloading my brain buffer. AFAIK I made the GC and node's destructor be able to run whatever is happening on other threads, making the whole thing threadsafe.

@emilio Do you think you have a bit of time to look at it? And maybe I should ping some other people, such as @bholley who initially filed #14213.

**Comment by emilio** [comment]
Yeah, I'll try to look at it soonish. It may be worth to hooking it into Gecko and give it a run through try so that we can catch any obvious problems. The refcount logging rule nodes have would also be something we should keep to ensure we don't leak them.

**Comment by nox** [comment]
Yeah I don't intend to make use of the crate directly in Stylo this was just for ease of hacking.

**Comment by nox** [comment]
This was fixed in #26227.
